datasets.py

```diff
@@ -31,47 +38,188 @@ def create_mnist_datasets(
     test_dataset = tf.data.Dataset.from_tensor_slices(test_images).shuffle(test_size).batch(batch_size)
     return train_dataset, test_dataset
 
+'''
```

Comment:
General comment: it'd be great to add `typing` and docstrings to all functions.

Reply:
yeah, agreed. I can do that once the dust settles

Reply:
(though I'm not actually sure how to indicate types for optional / conditional return values)